File: frontend/src/stores/tags.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import api from '@/lib/api'
import type { Tag, TagCreatePayload } from '@/types'

export const useTagsStore = defineStore('tags', () => {
  const tags = ref<Tag[]>([])
  const loading = ref(false)

  async function fetchTags() {
    loading.value = true
    try {
      const { data } = await api.get('/tags')
      tags.value = data
    } finally {
      loading.value = false
    }
  }

  async function createTag(payload: TagCreatePayload): Promise<Tag> {
    const { data } = await api.post('/tags', payload)
    await fetchTags()
    return data
  }

  async function deleteTag(id: string) {
    await api.delete(`/tags/${id}`)
    await fetchTags()
  }

  return {
    tags,
    loading,
    fetchTags,
    createTag,
    deleteTag,
  }
})

File: backend/app/api/tags.py
import uuid

from fastapi import APIRouter, HTTPException, status
from sqlalchemy import select

from app.api.deps import AuthToken, DbSession
from app.models import Tag
from app.schemas.tag import TagCreate, TagResponse, TagUpdate

router = APIRouter(prefix="/tags", tags=["tags"])


@router.get("", response_model=list[TagResponse])
async def list_tags(db: DbSession, _token: AuthToken):
    result = await db.execute(select(Tag).order_by(Tag.name))
    return result.scalars().all()


@router.post("", response_model=TagResponse, status_code=status.HTTP_201_CREATED)
async def create_tag(data: TagCreate, db: DbSession, _token: AuthToken):
    existing = await db.execute(select(Tag).where(Tag.name == data.name))
    if existing.scalar_one_or_none():
        raise HTTPException(status_code=409, detail="Tag already exists")

    tag = Tag(name=data.name, color=data.color)
    db.add(tag)
    await db.commit()
    await db.refresh(tag)
    return tag


@router.get("/{tag_id}", response_model=TagResponse)
async def get_tag(tag_id: uuid.UUID, db: DbSession, _token: AuthToken):
    tag = await db.get(Tag, tag_id)
    if not tag:
        raise HTTPException(status_code=404, detail="Tag not found")
    return tag


@router.put("/{tag_id}", response_model=TagResponse)
async def update_tag(tag_id: uuid.UUID, data: TagUpdate, db: DbSession, _token: AuthToken):
    tag = await db.get(Tag, tag_id)
    if not tag:
        raise HTTPException(status_code=404, detail="Tag not found")

    update_data = data.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(tag, field, value)

    await db.commit()
    await db.refresh(tag)
    return tag


@router.delete("/{tag_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_tag(tag_id: uuid.UUID, db: DbSession, _token: AuthToken):
    tag = await db.get(Tag, tag_id)
    if not tag:
        raise HTTPException(status_code=404, detail="Tag not found")

    await db.delete(tag)
    await db.commit()
